# Test cricket bowling analysis with Python

#### OBJECTIVES

- CSV file reading in different strategies
- Number of rows and columns in the dataframe
- Data statistics
- Checking missing values and data types
- Renaming the column names
- Removing a column from the dataframe
- Removing the columns BBI and BBM
- Number of players played for ICC
- Number of Australian Bowlers are present in this dataset
- Number of Bangladeshi Bowlers are present in this dataset
- Number of different countries are present in the dataset
- Players had played for the ligest period and shortest period of time
- Players who had the lowest economy rate
- Players who had the lowest strike rate
- Players who had the lowest bowling average

In [55]:
import pandas as pd
import numpy as np

# CSV file reading in different strategies

In [56]:
df = pd.read_csv("wickets.csv", encoding = "ISO-8859-1")

display(df.head())

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3

- Here I performed CSV file reading using pandas

In [57]:
df = pd.read_csv("wickets.csv", encoding = 'unicode_escape')

display(df.head())

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3

In [58]:
df = pd.read_csv("wickets.csv", engine = 'python')

display(df.head())

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3

In [59]:
df = pd.read_csv("wickets.csv", encoding = 'unicode_escape')

display(df.head(10))
print(df.shape)

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   
5          SCJ Broad (ENG)  2007-2021   149   274  29863  14590   524   8/15   
6            CA Walsh (WI)  1984-2001   132   242  30019  12688   519   7/37   
7            DW Steyn (SA)  2004-2019    93   171  18608  10077   439   7/51   
8      N Kapil Dev (INDIA)  1978-1994   131   227  27740  12867   434   9/83   
9        HMRKB Herath (SL)  1999-2018    93   170  25993  12157   433  9/127   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3  
5  11/121  27.84  2.93  56.9  18   3  
6   13/55  24.44  2.53  57.8  22   3  
7   11/60  22.95  3.24  42.3  26   5  
8  11/146  29.64  2.78  63.9  23   2  
9  14/184  28.07  2.80  60.0  34   9

(79, 14)


# Number of rows and columns in the dataframe

In [60]:
# Number of rows count
print("Total number of rows = ", df.shape[0])

# Number of columns count
print("Total number of columns = ", df.shape[1])

Total number of rows =  79
Total number of columns =  14


- In the dataset total number of rows and total number of columns are 79 and 14

# Data statistics

In [61]:
# Data statistics checking
display(df.describe())

Inns         Balls          Runs        Wkts        Ave  \
count   79.000000     79.000000     79.000000   79.000000  79.000000   
mean   144.911392  18638.354430   8599.354430  317.215190  27.469747   
std     51.180222   7199.256972   3085.168807  121.924911   3.655658   
min     67.000000   8785.000000   4846.000000  200.000000  20.940000   
25%    110.000000  13583.000000   6456.500000  229.000000  24.500000   
50%    129.000000  16498.000000   7742.000000  266.000000  28.000000   
75%    169.000000  21742.500000   9756.000000  374.500000  29.870000   
max    304.000000  44039.000000  18355.000000  800.000000  34.790000   

            Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  
mean    2.806835  59.193671  16.354430   2.797468  
std     0.351577   9.350132   9.642372   3.235935  
min     1.980000  41.200000   3.000000   0.000000  
25%     2.600000  53.300000   9.500000   1.000000  
50%     2.820000  57.400000  14.000000   2.000000  
75%     3.080000  63.950000  20.500000   3.500000  
max     3.460000  91.900000  67.000000  22.000000

- The earlier data statistics show that the minimum wickets taken by the bowler is 200 and the maximum wickets taken by the bowler is 800. As per the dataset, 25%, 50%, 75% of taken wickets are 229, 266, 374 respectively. The mean value and standard value are 317.21 and 121.92 in terms of mean and standard deviation.

# Checking missing values and data types

In [62]:
# Verifying each column for missing values and data types
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     object 
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(6), object(5)
memory usage: 8.8+ KB
None


- The above statistics show that there are no missing values because the total number of entries is 79, which is absolutely match with the statistics. Also, integer data types, objects, floats are sustaing the above statistics.

# Choosing columns by name from a csv file

In [63]:
col_names = ['Player', 'Mat', 'Inns', 'Wkts']
df_usecols = pd.read_csv("wickets.csv", encoding = 'unicode_escape', usecols=col_names)

display(df_usecols.head(5))
display(df_usecols.tail(5))

Player   Mat  Inns  Wkts
0  M Muralitharan (ICC/SL)   133   230   800
1           SK Warne (AUS)   145   273   708
2        JM Anderson (ENG)  164*   304   623
3         A Kumble (INDIA)   132   236   619
4         GD McGrath (AUS)   124   243   563

Player Mat  Inns  Wkts
74      SCG MacGill (AUS)  44    85   208
75  Saqlain Mushtaq (PAK)  49    86   208
76       AME Roberts (WI)  47    90   202
77          JA Snow (ENG)  49    93   202
78       JR Thomson (AUS)  51    90   200

# Renaming the column names

In [64]:
print(df.columns)

Index(['Player', 'Span', 'Mat', 'Inns', 'Balls', 'Runs', 'Wkts', 'BBI', 'BBM',
       'Ave', 'Econ', 'SR', '5', '10'],
      dtype='object')


In [65]:
df = df.rename(columns={'Mat':'Matches', 
                        'Inns':'Innings',
                        'Wkts': 'Wickets',
                        'Ave': 'Average',
                        'Econ': 'Economy',
                        'SR': 'Strike_Rate',
                        '5': 'Five_Wickets_Haul',
                        '10': 'Ten_Wickets_Haul'})

display(df.head())

Player       Span Matches  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010     133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007     145      273  40705  17995      708   
2        JM Anderson (ENG)  2003-2021    164*      304  35079  16575      623   
3         A Kumble (INDIA)  1990-2008     132      236  40850  18355      619   
4         GD McGrath (AUS)  1993-2007     124      243  29248  12186      563   

     BBI     BBM  Average  Economy  Strike_Rate  Five_Wickets_Haul  \
0   9/51  16/220    22.72     2.47         55.0                 67   
1   8/71  12/128    25.41     2.65         57.4                 37   
2   7/42   11/71    26.60     2.83         56.3                 30   
3  10/74  14/149    29.65     2.69         65.9                 35   
4   8/24   10/27    21.64     2.49         51.9                 29   

   Ten_Wickets_Haul  
0                22  
1                10  
2                 3  
3                 8  
4                 3

- Here, I renamed the columns as Matches, Innings, Wickets, Average, Economy, Strike_Rate, Five_Wickets_Haul, Ten_Wickets_Haul from Mat, Inns, Wkts, Ave, Econ, SR, 5, 10, respecively.

# Removing a column from the dataframe

In [66]:
df.drop('Balls', axis=1, inplace=True)

display(df.head())

Player       Span Matches  Innings   Runs  Wickets    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010     133      230  18180      800   9/51   
1           SK Warne (AUS)  1992-2007     145      273  17995      708   8/71   
2        JM Anderson (ENG)  2003-2021    164*      304  16575      623   7/42   
3         A Kumble (INDIA)  1990-2008     132      236  18355      619  10/74   
4         GD McGrath (AUS)  1993-2007     124      243  12186      563   8/24   

      BBM  Average  Economy  Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul  
0  16/220    22.72     2.47         55.0                 67                22  
1  12/128    25.41     2.65         57.4                 37                10  
2   11/71    26.60     2.83         56.3                 30                 3  
3  14/149    29.65     2.69         65.9                 35                 8  
4   10/27    21.64     2.49         51.9                 29                 3

- Here I removed the column names **Balls** from the previous dataframe.

# Removing the columns BBI and BBM

In [67]:
df.drop('BBI', axis=1, inplace=True)
df.drop('BBM', axis=1, inplace=True)

display(df.head())

Player       Span Matches  Innings   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010     133      230  18180      800   
1           SK Warne (AUS)  1992-2007     145      273  17995      708   
2        JM Anderson (ENG)  2003-2021    164*      304  16575      623   
3         A Kumble (INDIA)  1990-2008     132      236  18355      619   
4         GD McGrath (AUS)  1993-2007     124      243  12186      563   

   Average  Economy  Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul  
0    22.72     2.47         55.0                 67                22  
1    25.41     2.65         57.4                 37                10  
2    26.60     2.83         56.3                 30                 3  
3    29.65     2.69         65.9                 35                 8  
4    21.64     2.49         51.9                 29                 3

- Here I removed the column names **BBI** and **BBM** from the previous dataframe.

# Number of players played for ICC

In [76]:
file = df.copy()

display(file.head())

Span Matches  Innings   Runs  Wickets  Average  Economy  Strike_Rate  \
0  1992-2010     133      230  18180      800    22.72     2.47         55.0   
1  1992-2007     145      273  17995      708    25.41     2.65         57.4   
2  2003-2021    164*      304  16575      623    26.60     2.83         56.3   
3  1990-2008     132      236  18355      619    29.65     2.69         65.9   
4  1993-2007     124      243  12186      563    21.64     2.49         51.9   

   Five_Wickets_Haul  Ten_Wickets_Haul           Player Country  
0                 67                22  M Muralitharan   ICC/SL  
1                 37                10        SK Warne      AUS  
2                 30                 3     JM Anderson      ENG  
3                 35                 8        A Kumble    INDIA  
4                 29                 3      GD McGrath      AUS

In [77]:
# Rearranging the columns name
new_col_sequence = ['Player', 'Country', 'Matches', 'Innings', 'Wickets', 'Runs', 'Average',
                    'Economy', 'Strike_Rate', 'Five_Wickets_Haul', 'Ten_Wickets_Haul', 'Span']
df = df[new_col_sequence]

display(df.head())

Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
0  M Muralitharan   ICC/SL     133      230      800  18180    22.72     2.47   
1        SK Warne      AUS     145      273      708  17995    25.41     2.65   
2     JM Anderson      ENG    164*      304      623  16575    26.60     2.83   
3        A Kumble    INDIA     132      236      619  18355    29.65     2.69   
4      GD McGrath      AUS     124      243      563  12186    21.64     2.49   

   Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul       Span  
0         55.0                 67                22  1992-2010  
1         57.4                 37                10  1992-2007  
2         56.3                 30                 3  2003-2021  
3         65.9                 35                 8  1990-2008  
4         51.9                 29                 3  1993-2007

In [86]:
def icc_playing_check(xl):
    if "ICC" in xl:
        return "Yes"
    else:
        return "No"

In [87]:
df['played_for_ICC'] = df['Country'].apply(icc_playing_check)

display(df.head(5))

<ipython-input-87-c3e580ba2f02>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['played_for_ICC'] = df['Country'].apply(icc_playing_check)


Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
0  M Muralitharan   ICC/SL     133      230      800  18180    22.72     2.47   
1        SK Warne      AUS     145      273      708  17995    25.41     2.65   
2     JM Anderson      ENG    164*      304      623  16575    26.60     2.83   
3        A Kumble    INDIA     132      236      619  18355    29.65     2.69   
4      GD McGrath      AUS     124      243      563  12186    21.64     2.49   

   Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul       Span played_for_ICC  
0         55.0                 67                22  1992-2010            Yes  
1         57.4                 37                10  1992-2007             No  
2         56.3                 30                 3  2003-2021             No  
3         65.9                 35                 8  1990-2008             No  
4         51.9                 29                 3  1993-2007             No

In [88]:
df['played_for_ICC'].value_counts()

No     74
Yes     5
Name: played_for_ICC, dtype: int64

- The upwards data statistics showed that 5 players played for ICC and 74 players did not play for ICC.

 # Number of Australian Bowlers are present in this dataset

In [89]:
def Australian_bowler_check(xl):
    if "AUS" in xl:
        return "Yes"
    else:
        return "No"

In [90]:
df['played_for_Australia'] = df['Country'].apply(Australian_bowler_check)

display(df.head(5))

<ipython-input-90-ab519da138e5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['played_for_Australia'] = df['Country'].apply(Australian_bowler_check)


Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
0  M Muralitharan   ICC/SL     133      230      800  18180    22.72     2.47   
1        SK Warne      AUS     145      273      708  17995    25.41     2.65   
2     JM Anderson      ENG    164*      304      623  16575    26.60     2.83   
3        A Kumble    INDIA     132      236      619  18355    29.65     2.69   
4      GD McGrath      AUS     124      243      563  12186    21.64     2.49   

   Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul       Span played_for_ICC  \
0         55.0                 67                22  1992-2010            Yes   
1         57.4                 37                10  1992-2007             No   
2         56.3                 30                 3  2003-2021             No   
3         65.9                 35                 8  1990-2008             No   
4         51.9                 29                 3  1993-2007             No   

  played_for_Australia  
0                   No  
1                  Yes  
2                   No  
3                   No  
4                  Yes

In [93]:
df['played_for_Australia'].value_counts()

No     61
Yes    18
Name: played_for_Australia, dtype: int64

- The earlier results showed that 18 Australian players participated in test cricket, according to the dataset.

 # Number of Bangladeshi Bowlers are present in this dataset

In [101]:
def Bangladeshi_bowler_check(xl):
    if "BDESH" in xl:
        return "Yes"
    else:
        return "No"

In [102]:
df['played_for_Bangladesh'] = df['Country'].apply(Bangladeshi_bowler_check)

display(df.head(5))

<ipython-input-102-cf41bb29e476>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['played_for_Bangladesh'] = df['Country'].apply(Bangladeshi_bowler_check)


Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
0  M Muralitharan   ICC/SL     133      230      800  18180    22.72     2.47   
1        SK Warne      AUS     145      273      708  17995    25.41     2.65   
2     JM Anderson      ENG    164*      304      623  16575    26.60     2.83   
3        A Kumble    INDIA     132      236      619  18355    29.65     2.69   
4      GD McGrath      AUS     124      243      563  12186    21.64     2.49   

   Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul       Span played_for_ICC  \
0         55.0                 67                22  1992-2010            Yes   
1         57.4                 37                10  1992-2007             No   
2         56.3                 30                 3  2003-2021             No   
3         65.9                 35                 8  1990-2008             No   
4         51.9                 29                 3  1993-2007             No   

  played_for_Australia played_for_Bangladesh  
0                   No                    No  
1                  Yes                    No  
2                   No                    No  
3                   No                    No  
4                  Yes                    No

In [104]:
df['played_for_Bangladesh'].value_counts()

No     78
Yes     1
Name: played_for_Bangladesh, dtype: int64

- The above statistics show that one Bangladeshi is sustained in the list.

# Number of different countries are present in the dataset

In [100]:
df['Country'].value_counts()

AUS        18
ENG        13
INDIA      10
WI          9
SA          7
PAK         7
NZ          6
SL          2
ENG/ICC     2
ICC/SL      1
ICC/SA      1
ZIM         1
BDESH       1
ICC/NZ      1
Name: Country, dtype: int64

- From the above list, we saw that 10 different countries' players participated in test cricket.
- Also, four different nations named ENG, SL, SA, and NZ players participated as ICC participants.

In [105]:
# Span column splitting depends on the "-"
df_span = df['Span'].str.split("-", expand=True)

# conccatination the new data frame with the old data frame
df = pd.concat([df, df_span], axis=1)

# renaming the newly created column names
df = df.rename(columns={0: "start_year",
                        1: "end_year"})

# removing the span column
df = df.drop("Span", axis=1)

display(df.head())

Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
0  M Muralitharan   ICC/SL     133      230      800  18180    22.72     2.47   
1        SK Warne      AUS     145      273      708  17995    25.41     2.65   
2     JM Anderson      ENG    164*      304      623  16575    26.60     2.83   
3        A Kumble    INDIA     132      236      619  18355    29.65     2.69   
4      GD McGrath      AUS     124      243      563  12186    21.64     2.49   

   Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
0         55.0                 67                22            Yes   
1         57.4                 37                10             No   
2         56.3                 30                 3             No   
3         65.9                 35                 8             No   
4         51.9                 29                 3             No   

  played_for_Australia played_for_Bangladesh start_year end_year  
0                   No                    No       1992     2010  
1                  Yes                    No       1992     2007  
2                   No                    No       2003     2021  
3                   No                    No       1990     2008  
4                  Yes                    No       1993     2007

In [ ]:
df['start_year'] = df['start_year'].astype('int') 
df['end_year'] = df['end_year'].astype('int')

print(df.info())

# Players who had played for the longest period and shortest period of time

In [111]:
df.sort_values(by='years_played', ascending = False).head(5)

Player Country Matches  Innings  Wickets   Runs  Average  \
21      Imran Khan      PAK      88      142      362   8258    22.81   
55       GS Sobers       WI      93      159      235   7999    34.03   
9     HMRKB Herath       SL      93      170      433  12157    28.07   
0   M Muralitharan   ICC/SL     133      230      800  18180    22.72   
2      JM Anderson      ENG    164*      304      623  16575    26.60   

    Economy  Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
21     2.54         53.7                 23                 6             No   
55     2.22         91.9                  6                 0             No   
9      2.80         60.0                 34                 9             No   
0      2.47         55.0                 67                22            Yes   
2      2.83         56.3                 30                 3             No   

   played_for_Australia played_for_Bangladesh  years_played  
21                   No                    No            21  
55                   No                    No            20  
9                    No                    No            19  
0                    No                    No            18  
2                    No                    No            18

- From the above statistics, we saw that a player named Imran Khan played test cricket for the longest time, 21 years.
- The player named GS Sobers played test cricket for 20 years in the second longest time period.

In [112]:
df.sort_values(by='years_played', ascending = False).tail(5)

Player  Country Matches  Innings  Wickets  Runs  Average  Economy  \
54    Yasir Shah       PAK     46*       84      235  7248    30.84     3.19   
61   SJ Harmison   ENG/ICC      63      115      226  7192    31.82     3.22   
72  JR Hazlewood       AUS      55      103      212  5438    25.65     2.74   
71      K Rabada        SA      47       86      213  4846    22.75     3.30   
44      GP Swann       ENG      60      109      255  7642    29.96     2.98   

    Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
54         57.9                 16                 3             No   
61         59.1                  8                 1            Yes   
72         56.0                  9                 0             No   
71         41.2                 10                 4             No   
44         60.1                 17                 3             No   

   played_for_Australia played_for_Bangladesh  years_played  
54                   No                    No             7  
61                   No                    No             7  
72                  Yes                    No             7  
71                   No                    No             6  
44                   No                    No             5

- From the above statistics, we saw that a player named GP Swann played test cricket for the shortest time, 5 years.
- The player named K Rabada played test cricket for 6 years in the second shortest time period.

# Players who had the lowest economy rate

In [120]:
df.sort_values(by='Economy', ascending = False).tail(5)

Player Country Matches  Innings  Wickets  Runs  Average  Economy  \
68   CV Grimmett      AUS      37       67      216  5231    24.21     2.16   
39       BS Bedi    INDIA      67      118      266  7637    28.71     2.14   
47      R Benaud      AUS      63      116      248  6704    27.03     2.10   
35  DL Underwood      ENG      86      151      297  7674    25.83     2.10   
32      LR Gibbs       WI      79      148      309  8989    29.09     1.98   

    Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
68         67.1                 21                 7             No   
39         80.3                 14                 1             No   
47         77.0                 16                 1             No   
35         73.6                 17                 6             No   
32         87.7                 18                 2             No   

   played_for_Australia played_for_Bangladesh  years_played  
68                  Yes                    No            11  
39                   No                    No            13  
47                  Yes                    No            12  
35                   No                    No            16  
32                   No                    No            18

- From the above statistics, we saw that the lowest economic rate was 1.98 for a player named LR Gibbs.
- The second lowest economic rate is 2.10, which is jointly carried by DL Underwood and R Benaud.

# Players who had the lowest strike rate

In [121]:
df.sort_values(by='Strike_Rate', ascending = False).tail(5)

Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
25     AA Donald       SA      72      129      330   7344    22.25     2.83   
19   MD Marshall       WI      81      151      376   7876    20.94     2.68   
20  Waqar Younis      PAK      87      154      373   8788    23.56     3.25   
7       DW Steyn       SA      93      171      439  10077    22.95     3.24   
71      K Rabada       SA      47       86      213   4846    22.75     3.30   

    Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
25         47.0                 20                 3             No   
19         46.7                 22                 4             No   
20         43.4                 22                 5             No   
7          42.3                 26                 5             No   
71         41.2                 10                 4             No   

   played_for_Australia played_for_Bangladesh  years_played  
25                   No                    No            10  
19                   No                    No            13  
20                   No                    No            14  
7                    No                    No            15  
71                   No                    No             6

- From the above statistics, we saw that the lowest strike rate was 41.2 for a player named K Rabada.
- The second lowest strike rate is 42.3, which is bearing DW. Steyn.

# Players who had the lowest bowling average

In [122]:
df.sort_values(by='Average', ascending = False).tail(5)

Player Country Matches  Innings  Wickets   Runs  Average  Economy  \
4    GD McGrath      AUS     124      243      563  12186    21.64     2.49   
33   FS Trueman      ENG      67      127      307   6625    21.57     2.61   
15  CEL Ambrose       WI      98      179      405   8501    20.99     2.30   
41     J Garner       WI      58      111      259   5433    20.97     2.47   
19  MD Marshall       WI      81      151      376   7876    20.94     2.68   

    Strike_Rate  Five_Wickets_Haul  Ten_Wickets_Haul played_for_ICC  \
4          51.9                 29                 3             No   
33         49.4                 17                 3             No   
15         54.5                 22                 3             No   
41         50.8                  7                 0             No   
19         46.7                 22                 4             No   

   played_for_Australia played_for_Bangladesh  years_played  
4                   Yes                    No            14  
33                   No                    No            13  
15                   No                    No            12  
41                   No                    No            10  
19                   No                    No            13

- From the above statistics, we saw that the lowest bowling average was 20.94 for a player named MD Marshall. 
- The second lowest bowling average is 20.97, which is by J Garner.